raft-tech/df-arcadedb
# Arcade History and Rollback, along with saving metadata of changes

# Overview

> _Describe your changes.  This can also be the commit message (if it was a [good one](https://cbea.ms/git-commit/))._

## Related Issues

> _Reference any issue(s) this PR should either close or is related to in some way._

* Closes raft-tech/data-fabric#2853

# Test Procedure

## Setup

From a new shell session in any clean/empty directory, perform the following 👇.

### Environment

1. Export your GitHub credentials to pull from the GitHub container registry.
    ```shell
    export GHP_USERNAME=your_github_username_goes_here
    export GHP_SECRET=your_github_pat_goes_here
    ```

1. Create a subdirectory for this PR test.
    ```shell
    export PR_HOME=$(pwd)/pr-test
    mkdir -pv $PR_HOME
    ```

### Clone Repositories

1. Clone the main `data-fabric` repo, checking out the branch we'll use to test this PR.
    ```shell
    export DF_HOME=$PR_HOME/data-fabric
    git clone --branch dev https://github.com/raft-tech/data-fabric.git $DF_HOME
    ```

Related DF PR https://github.com/raft-tech/data-fabric/pull/3050

1. Clone the `df-arcadedb`, checking out the branch for this PR.
    ```shell
    export DF_ARCADEDB_HOME=$PR_HOME/df-arcadedb    
    git clone --branch YOUR_PR_BRANCH https://github.com/raft-tech/df-arcadedb.git $DF_ARCADEDB_HOME
    ```

1. Clone `dfdev` to help with building and running things. \
   ℹ️ _If you already have [`dfdev`](https://github.com/raft-tech/dfdev) you can skip this step._
    ```shell
    git clone https://github.com/raft-tech/dfdev.git $PR_HOME/dfdev
    PATH=$PATH:$PR_HOME/dfdev/dfdev
    ```
   
### Deploy

1. (Re)Create a new local `kind` cluster and deploy Data Fabric to it.
    ```shell
    dfdev cluster recreate
    ```

1. Re-build and re-deploy `df-arcadedb`

    ```shell
    dfdev redeploy df-arcadedb
    ``` 

## Verification Steps

1. Log into arcadedb.localhost
2. Sign in with keycloak admin user and password (no SSO yet)
3. Create a new database, call it alpha
4. Switch to the new database in the dropdown
5. Switch to the first tab for running SQL
6. `CREATE VERTEX TYPE People`
7. 
    ```
    INSERT INTO People CONTENT 
    {
      firstName: 'Enzo',
      lastName: 'Ferrari'
    }
    ```
8. Confirm the record inserts correctly with created and createdBy fields set. `select from `People` limit 30`
9. `UPDATE People set firstName='a'`
10. Confirm the record updates correctly with lastModified and lastModifiedBy fields set. `select from `People` limit 30`
11. In df-frontend, navigate to the streaming window
12. Click to enable CDC Stream for the `arcadedbcdc--admin--alpha` topic. (Clunky I know, CDC streaming was supposed to be toggleable from arcade, not streaming to kafka by default and optionally dumping to delta)
13. Wait a few minutes...
14. In df-frontend, navigate to the swagger docs for arcade. Authorize with admin creds
15. Scroll down to `/history/{database}/{entityType}/{id}` and click try out
16. Fill in `database=alpha`, `entityType=People`, `id=1:0`.
17. Click execute. You ought to get a successful response with 2 history objects. If you get an error, try giving it a few more minutes. You can open k9s and check the status of the cdc job in the `arcadedbcdc--admin--alpha` pod
18. Grab the eventId for the historical record where the object was created in the results
19. Go down to the next, rollback operation in the arcade swagger.
20. Fill in same options as before, additionally adding `eventId=[the event id you grabbed]`
21. Click execute. You should get a successful response
22. Navigate back to arcade
23. Run the query `select from `People` limit 30`
24. Confirm the record rolled back. First name should be `Enzo`. The last modified date should've updated to right now when you ran the rollback.
